File: sword-means-offer-2/005/main.go
package main

import "fmt"

// @Time     : 2021/12/14 22:34
// @Author   : Ranshi
// @File     : main.go
// @Doc      : 剑指 Offer II 005. 单词长度的最大乘积

func maxInt(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func maxProduct(words []string) (ans int) {
	n := len(words)
	masksMap := make(map[int]int)
	for i := 0; i < n; i++ {
		bitMask := 0
		for _, c := range words[i] {
			bitMask |= (1 << (c - 'a'))
		}
		if len(words[i]) > masksMap[bitMask] {
			masksMap[bitMask] = len(words[i])
		}
	}

	for x := range masksMap {
		for y := range masksMap {
			if (x & y) == 0 {
				length := masksMap[x] * masksMap[y]
				ans = maxInt(length, ans)
			}
		}
	}
	return
}

func main() {
	fmt.Println(maxProduct([]string{"abcw", "baz", "foo", "bar", "fxyz", "abcdef"}))
}

File: sword-means-offer-2/005/cpp/main.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
  int maxProduct(vector<string> &words) {
    int length = words.size();
    vector<int> masks(length);
    for (size_t i = 0; i < length; i++) {
      string word = words[i];
      int word_length = word.size();
      for (size_t j = 0; j < word_length; j++) {
        masks[i] |= 1 << (word[j] - 'a');
      }
    }
    int max_prod = 0;
    for (size_t i = 0; i < length; i++) {
      for (int j = i + 1; j < length; j++) {
        if ((masks[i] & masks[j]) == 0) {
          max_prod = max(max_prod, int(words[i].size() * words[j].size()));
        }
      }
    }
    return max_prod;
  }
};

int main(int argc, char const *argv[]) {
  vector<string> words = {"abcw", "baz", "foo", "bar", "fxyz", "abcdef"};
  auto res = Solution().maxProduct(words);
  cout << res;
  return 0;
}
